kratos/solving_strategies/schemes/residual_based_pseudo_static_displacement_scheme.h

```diff
@@ -153,52 +154,76 @@ class ResidualBasedPseudoStaticDisplacementScheme
     void Predict(
         ModelPart& rModelPart,
         DofsArrayType& rDofSet,
-        TSystemMatrixType& A,
-        TSystemVectorType& Dx,
-        TSystemVectorType& b
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb
         ) override
     {
         KRATOS_TRY;
 
-        const double delta_time = rModelPart.GetProcessInfo()[DELTA_TIME];
+        // Process info
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        // Delta time
+        const double delta_time = r_current_process_info[DELTA_TIME];
 
         // Updating time derivatives (nodally for efficiency)
+        const auto it_node_begin = rModelPart.Nodes().begin();
         const int num_nodes = static_cast<int>(rModelPart.NumberOfNodes());
 
-        array_1d<double, 3> zero_array(3, 0.0);
-        array_1d<double, 3> delta_displacement;
+        array_1d<double, 3> zero_array = ZeroVector(3);
+        array_1d<double, 3> delta_displacement = zero_array;
 
         #pragma omp parallel for private(zero_array, delta_displacement)
         for(int i = 0;  i < num_nodes; ++i) {
-            auto it_node = rModelPart.Nodes().begin() + i;
+            auto it_node = it_node_begin + i;
 
-            //Predicting: NewDisplacement = previous_displacement + previous_velocity * delta_time;
+            //Predicting: r_current_displacement = r_previous_displacement + r_previous_velocity * delta_time;
             //ATTENTION::: the prediction is performed only on free nodes
 
-            const array_1d<double, 3>& previous_velocity     = (it_node)->FastGetSolutionStepValue(VELOCITY,     1);
-            const array_1d<double, 3>& previous_displacement = (it_node)->FastGetSolutionStepValue(DISPLACEMENT, 1);
-            array_1d<double, 3>& current_acceleration        = (it_node)->FastGetSolutionStepValue(ACCELERATION);
-            array_1d<double, 3>& current_velocity            = (it_node)->FastGetSolutionStepValue(VELOCITY);
-            array_1d<double, 3>& current_displacement        = (it_node)->FastGetSolutionStepValue(DISPLACEMENT);
-
-            if (it_node -> IsFixed(DISPLACEMENT_X) == false)
-                current_displacement[0] = previous_displacement[0] + delta_time * previous_velocity[0];
+            const array_1d<double, 3>& r_previous_velocity     = (it_node)->FastGetSolutionStepValue(VELOCITY,     1);
+            const array_1d<double, 3>& r_previous_displacement = (it_node)->FastGetSolutionStepValue(DISPLACEMENT, 1);
+            array_1d<double, 3>& r_current_acceleration        = (it_node)->FastGetSolutionStepValue(ACCELERATION);
+            array_1d<double, 3>& r_current_velocity            = (it_node)->FastGetSolutionStepValue(VELOCITY);
+            array_1d<double, 3>& r_current_displacement        = (it_node)->FastGetSolutionStepValue(DISPLACEMENT);
+
+            if (it_node->IsFixed(VELOCITY_X)) {
```

Comment:
note that you cannot force both displacement and velocity.
This means that if the VELOCITY is fixed than the DISPLACEMENT is a consequence of it (or the contrary)

Saying it in another way, you need to choose who rules (for example prescribed VELOCITY) and live with that, in other words understand that if you prescribe the velocity the displacement will be a funtion of it

Reply:
@RiccardoRossi the problem is the following. This method is used on the Update and the PRedict. With the current changes the RPedict works fine and can be removed, with the update don't, because the Dx comes from the linear system solution, and you cannot control that

Reply:
@loumalouomega this is precisely the wrong argument :-)

you control completely dx on the fixed nodes! in fact dx=0 on the fixed nodes

Reply:
@RiccardoRossi in #4426 @salvalatorre fixes the velocity and not the displacements. Your approach will not work with the problem that @salvalatorre is trying to fix. Which problem do you have with checking the fixity anyway?

Reply:
that's the error. When velocity is fixed also displacement should be fixed. 
Note that if the displacement is not fixed, than the information will simply not be in the system as a Dirichlet conditon, so this will never work correctly.

I believe that this should be accomplished by having a specific process for setting First or Second derivatives, which takes care of fixing both.

An alternative that we discussed many months back was to have such a process to compute internally the displacement corresponding to the target velocity. 
The problem with this approach is that the calculation of the exact displacement increment that is consistent with a prescribed velocity will depend on the time scheme being employed, an information that is not available to the scheme. 
This was the rationale for  choosing to perform the calculation of the consistent displacements in the scheme predict instead of having it in the process.